\abstract{\bootranges provides fast functions for generation of background feature set to explore the null hypothesis of enrichment analysis, colocalization or connect cis-regulatory elements (cCREs) pairs. Result shows that the conventional bootstrap or permutation can have higher significance which may results low specificity in some cases. In addition, \bootranges provides statistical reasonable and flexible effect size cut-offs at different experiment designs, eg. \textit{p}value in GWAS, logFC in differential expression gene (DEG) analysis.
The \bootranges is integrated into the an R/Bioconductor package \emph{nullranges}
at \url{https://bioconductor.org/packages/nullranges}.\\
% \textbf{Supplementary information:} Supplementary data are available at Bioinformatics online.
%\textbf{Contact:} \href{michaelisaiahlove@gmail.com}{michaelisaiahlove@gmail.com}
}