\section{Introduction}

In genomic analysis, to assess whether
there is an enrichment of one set of features near another 
one must choose an appropriate null model \citep{reviewdilemma2014}.
For example, an enrichment of ATAC-seq peaks near certain genes
may indicate a regulatory relationship \citep{lee2020fluent}, 
and enrichment of GWAS SNPs near tissue-specific ATAC-seq peaks may
suggest mechanisms underlying the GWAS trait.
Such analyses rely on specifying a null distribution, where one
strategy is to uniformly shuffle one set of the
genomic features in the genome, possibly while considering a set of
excluded regions.
However, uniformly distributed null feature sets will not exhibit the
clumping property common with genomic feature sets.
Using an overly simplistic null distribution that doesn't take into
account local dependencies could result in misleading conclusions.
More sophisticated methods exist, for example
GAT, which allows for controlling by local GC content
\citep{GAT_2013}, and regioneR, which implements a circular shift to
preserve clumping \citep{gel2016regioner}.
The block bootstrap \citep{politis1999subsampling}
provides an alternative, where one instead generates
random feature sets by sampling large blocks of features from the
original set with replacement, as proposed for 
genomic features by \citet{bickel2010subsampling} in GSC.
Using the block bootstrap is more
computationally intensive than simple shuffling, and so GSC implements
a strategy of swapping pairs of blocks to compute overlaps, while
avoiding a genome-scale bootstrap.

Here we describe the \bootranges software, with efficient
vectorized code for performing block bootstrap sampling of
\granges \citep{lawrence2013software} objects.
\bootranges is part of a modular analysis workflow, where bootstrapped
\granges can be analyzed at block or genome scale using tidy
analysis with \plyranges \citep{lee2019plyranges}.
We provide recommendations for genome segmentation and block length
motivated by analysis of example datasets.
We demonstrate how \bootranges can be incorporated into complex
downstream analyses, including to motivate thresholds chosen during
enrichment analysis and single-cell multi-omics.

\vspace*{-20pt}

\section{Features}

\bootranges offers a simple ``unsegmented'' block bootstrap as well as
a ``segmented'' block bootstrap:
since the distribution of features in the genome exhibits multi-scale
structure, we follow the logic of \citet{bickel2010subsampling} and consider to
perform block bootstrapping within \textit{segments} of the genome, which are
more homogeneous in terms of base composition and feature density.
We consider various genome segmentation procedures based on gene
density, or annotations, e.g. Giemsa bands or pre-computed segments
(see software vignette for details).
The genome segments define large (e.g. on the order of ${\sim}1$ Mb),
relatively homogeneous segments within which to sample blocks
(\cref{fig:framework}A). 
The input for the workflow is \granges features \bm{$x$} and
\bm{$y$}, with optional metadata columns (\texttt{mcols}) that can be
used for computing a more complex test statistic than overlaps.
Given a segmentation and \texttt{blockLength} $L_b$, a \bootranges
object is generated, which concatenates \granges across bootstrap
iterations. This \bootranges object can be manipulated with \plyranges
to derive the bootstrap distribution of test statistics $\{s_r\}$, and a
bootstrap p-value for observed test statistic $s$:
$ \frac{1}{R} \sum_{r=1}^R 1_{\{s_r > s\}} $ (\cref{fig:framework}B).
The \bootranges algorithms are explained schematically in Supplementary \cref{sec:algorithm}.

\vspace{-0.5cm}
\begin{figure}[htbp]
\centering% default with `floatrow`
\setlength{\abovecaptionskip}{-0.05cm}
\includegraphics[scale=0.65]{Figures/bootRanges.jpg}
\caption{Overview of \bootranges workflow. (a) A schematic
  diagram of \bootranges with \texttt{blockLength} $L_b$ across chromosome.
  (b) Testing overlaps of features in \bm{$x$} with features in
  \bm{$y$}, and comparing to a bootstrap distribution.} 
\label{fig:framework}
\vspace{-0.5cm}
\end{figure}

\vspace*{-21pt}
\section{Application}

We first applied \bootranges to determine the significance of the
overlap of ATAC-seq on human liver tissue
\citep{CURRIN20211169} with SNPs associated with total cholesterol,
bootstrapping the SNPs to assess significance.
While the observed overlap was significant across many combinations of
various segmentation methods and block length according to empirical p-value, 
the variance of the
bootstrap statistics distribution and the resulting $z$ score varied greatly
(\cref{fig:result}A-B).
We used the $z$ score to measure the distance between the observed
value and bootstrap distribution in terms of standard deviations.
Overlap rate was defined as the proportion of
SNPs that had peaks within 10kb.
That the variance of the distribution in \cref{fig:result}A for the
unsegmented bootstrap increased with $L_b$ indicated that
bootstrapping with respect to a genome
segmentation may be a more appropriate choice
\citep{bickel2010subsampling}. 
% inhomogeneous and segmentation could alleviate the scenario.
The decreasing trend using pre-defined segmentation from
Roadmap Epigenomics indicated too many short segments,
where $L_b$ is too close to $L_s$ for effective block randomization.
To choose an optimal block length and segmentation, 
we considered a number of diagnostics including
those recommended previously \citep{bickel2010subsampling}:
the variance of the bootstrap distribution (\cref{fig:result}A),
a scaled version of the change in the width of the previous block length
bootstrap distribution over $L_b$,
and the inter-feature distance distribution to assess conserved
clustering of features (Supplementary \cref{sec:length}).
Here $L_s \approx 2Mb$ and $L_b \in [300\textrm{kb},600\textrm{kb}]$ was 
shown to be a good range for segmentation and block
lengths (\cref{fig:suppfig}A-C, and Supplementary \cref{sec:results}).
The scientific conclusion of this example was that liver ATAC-seq were
much closer to total cholesterol SNPs than expected even when placing
blocks of SNPs to match a genome segmentation.
Shuffling of features (Supplementary \cref{sec:shuffle})
resulted in a much higher $z = 18.5$, compared to $z \approx 4$ for
the methods that preserve local correlation and place blocks with
respect to a genome segmentation.

%\vspace{-0.5cm}
\begin{figure}[hbtp]
\centering% default with `floatrow`
\setlength{\abovecaptionskip}{-0.1cm}
\setlength{\belowcaptionskip}{-0.1cm}
\includegraphics[scale=0.28]{Figures/fig2_3.jpeg}
\caption{
  Bootstrapping parameter evaluation and enrichment analyses. 
  A) Variance of the rate of overlaps and
  B) $z$ score for the observed overlap,
  for different segmentations and $L_b$ on the liver ATAT-seq
  dataset.
  %Upper and lower horizontal line represents the $z$ score obtained
  %with naive shuffling and ....
   C)The GAM predicted curves for observed (black line) and
  bootstrapped data (densities),
  for the overlap count over logFC of DE genes for the macrophage
  dataset.
  Color of density represents the $z$ score for the observed overlap
  with respect to the conditional density.
}
\label{fig:result}
\vspace{-0.7cm}
\end{figure}

%% z score, independent of number of bootstraps, was used to measure the
%% distance between the expected value and the observed one according to
%% the standard deviations.
%% the $z = 4.10$ if look at circular binary search(CBS)
%% \citep{cbs} segmentation method with $L_s = 2e6$ and
%% $L_b=5e5$(\cref{fig:result}B).
%% As seen in applications of \citet{bickel2010subsampling}, the effect
%% of segmentation did not greatly alter conclusions, e.g. rejection of
%% the null hypothesis, in this case, although the z score varies greatly
%% among the different segmentations and block lengths.

We demonstrate using \bootranges to motivate the choice of data-driven thresholds 
during enrichment analyses. We tested this on a dataset of differential chromatin accessibility and gene expression 
\citep{alasoo2018shared,lee2020fluent}.
A generalized linear model (GLM) with penalized splines was
fit to the overlap count over gene logFC, both for the original
data and to each of the bootstrap datasets.
Conditional densities of splines fit to bootstrap data
were computed at various thresholds to reveal how
the threshold choice would affect the
variance of the bootstrap density and the resulting $z$ score
(\cref{fig:result}C).
These analyses suggested that $|\textrm{logFC}| = 2$
are optimized thresholds where the $z$ score was highest
(\cref{fig:suppfig}E).

%We demonstrate using \bootranges to motivate the choice of thresholds 
%that are applied to feature sets during enrichment analyses.
%We tested this on the aforementioned liver ATAC-seq example, and on a
%dataset of differential chromatin accessibility and gene expression 
%\citep{alasoo2018shared,lee2020fluent}.
%A generalized linear model (GLM) with penalized regression splines was
%fit to the overlap rate or count over the $-\log_{10}(p)$ or
%gene logFC, both for the original
%data and to each of the bootstrap datasets.
%Condition densities of splines fit to bootstrap data
%were computed at various thresholds to reveal how
%the threshold choice would affect the
%variance of the bootstrap density and the resulting $z$ score
%(\cref{fig:result}C-D).
%These analyses suggested that $-\log_{10}(p) = 8$ and
%$|\textrm{logFC}| = 2$
%are optimized thresholds where the $z$ score was highest
%(\cref{fig:suppfig}D-E).

%We found that the $z$ score was highest when $-\log_{10}(p) = 8$
%(\cref{fig:suppfig}D),
%and when $\textrm{|logFC|} = 2$(\cref{fig:suppfig}E).
%(\cref{fig:suppfig}E).
%% from \textit{gam} function in the
%% \emph{mgcv} R package were fitted and \textit{predict\_gam} function
%% in the \emph{tidymv} R package were predicted on observed and each
%% null feature sets.
%% $$
%% \setlength{\abovedisplayskip}{3pt}
%% \setlength{\belowdisplayskip}{3pt}
%% log \left( \frac{\pi}{1-\pi} \right) = \beta_0  + f (-log_{10}p), log(\mu) = \beta_0 + f (log_{FC})
%% $$ 
%% for rate and count-based statistic, separately.
%% All generated 95\%
%% percentile intervals at the same time across a range of effect sizes
%% were displayed by conditional density plot 

We additionally applied \bootranges to Chromium Single Cell Multiome
ATAC- and RNA-seq, to assess the correlation ($\rho$) of log counts for the two
modalities for all pairs of genes and peaks, across
14 cell types (pseudo-bulk). Across all genes, we observed
$\bar{\rho} = 0.33$, which was 
significantly higher than the bootstrap correlation mean
(\cref{fig:suppfig}F, $\bar{\rho}_{R} = 0.007$). Thus, as expected, RNA
and ATAC measured at local peaks had similar cell-type-specificity.
Additionally, the average bootstrapped gene-peaks correlation per gene can be computed to
idenitfy gene-promoter pairs that are significantly correlated across cell types (\cref{fig:suppfig}G-H).
%Additionally, the average gene-promoter
%correlation per gene can be computed. 
%XXX of genes had a
%significantly higher correlation than bootstrapped data.

\vspace*{-20pt}
\section{Conclusion}
Overall, \bootranges generates null sets preserving 
genome clumping property and therefore derives more acurate conclusion in genomic analysis. 
It has great flexibility in variaous disciplines(e.g. identify putative transcription factor binding site according to enriched peak regions)
 and fast implementation time. We compared \bootranges and GSC speed using
ENCODE kidney and bladder ChIP-seq. The average time to
block bootstrap the genome using \bootranges was 0.30s and
0.37s adding overlap computation. A comparable
analysis with GSC took 7.56s.
% -- maybe we can mention the code link on page 1 to save space ...
% All of the R code and data used in this paper are available at the
% following repository: 
% \url{https://github.com/Wancen/bootRangespaper}.

\vspace*{-25pt}

\section*{Funding}
This work was funded by CZI EOSS and NIH [NHGRI R01 HG009937]. 

\vspace*{-25pt}
