\documentclass{article}

\usepackage{authblk}
\title{Supplementary Methods, Tables, and Figures}
 \author[1]{Wancen Mu}
 \author[2]{Eric Davis}
 \author[5]{Stuart Lee}
 \author[6]{Mikhail Dozmorov}
 \author[2,3]{Douglas H. Phanstiel}
 \author[1,4]{Michael I. Love \thanks{michaelisaiahlove@gmail.com}}
 \affil[1]{Department of Biostatistics, }
 \affil[2]{Curriculum in Bioinformatics and Computational Biology, }
 \affil[3]{Thurston Arthritis Research Center, Department of Cell Biology \& Physiology, Lineberger Comprehensive Cancer Center, Curriculum in Genetics \& Molecular Biology, and}
 \affil[4]{Department of Genetics, University of North Carolina-Chapel Hill, NC 27599}
 \affil[5]{Genentech, South San Francisco, CA, USA}
 \affil[6]{Department of Biostatistics, Department of Pathology, Virginia Commonwealth University, Richmond, VA 23298, USA}

\date{\today}
\usepackage{graphicx}
\usepackage{graphics}
\usepackage{amsmath}
\usepackage{threeparttable}
\usepackage[top=1in,left=1in,right=1in,bottom=1in]{geometry}
\usepackage[usenames,dvipsnames,svgnames,table]{xcolor}
\usepackage{xspace}
% \usepackage{hyperref}
\usepackage[colorlinks=true,linkcolor=black,citecolor=blue,urlcolor=blue,]{hyperref}
\usepackage{natbib}
\usepackage{xcite}
\usepackage{courier}

% cross-ref with main paper
\usepackage{xr}
\makeatletter
\newcommand*{\addFileDependency}[1]{% argument=file name and extension
  \typeout{(#1)}
  \@addtofilelist{#1}
  \IfFileExists{#1}{}{\typeout{No file #1.}}
}
\makeatother
\newcommand*{\myexternaldocument}[1]{%
    \externaldocument{#1}%
    \addFileDependency{#1.tex}%
    \addFileDependency{#1.aux}%
}
\myexternaldocument{plain}


\usepackage[ruled,vlined,linesnumbered,noresetcount]{algorithm2e}
\newcommand\mycommfont[1]{\footnotesize\ttfamily\textcolor{blue}{#1}}
\SetCommentSty{mycommfont}
\usepackage{float}
\usepackage[section]{placeins}
\usepackage[capitalise,noabbrev]{cleveref}
\usepackage{stfloats}
\usepackage{bm}
\usepackage{booktabs}
\usepackage{diagbox} % for diag line in table
\usepackage{hyperref}
\usepackage{geometry}
\usepackage{bm}
\usepackage{amsfonts,amssymb} %font
\usepackage{amsthm}
\usepackage{booktabs}
\usepackage{subfigure}
\geometry{a4paper,scale=0.8}
\renewcommand{\thefigure}{S\arabic{figure}}
\renewcommand{\thetable}{S\arabic{table}}
\renewcommand{\thesubfigure}{\Alph{subfigure}}

\usepackage[utf8x]{inputenc} % deal with "-" not usually set up by the [utf8] option

\input{wancen_commands}

\usepackage{listings}
\lstset{basicstyle=\footnotesize\ttfamily}
\lstset{numbers=left, %设置行号位置
        numberstyle=\tiny, %设置行号大小
        keywordstyle=\color{blue}, %设置关键字颜色
        commentstyle=\color[cmyk]{1,0,1,0}, %设置注释颜色
        frame=single, %设置边框格式
        escapeinside=``, %逃逸字符(1左面的键)，用于显示中文
        %breaklines, %自动折行
        extendedchars=false, %解决代码跨页时，章节标题，页眉等汉字不显示的问题
        xleftmargin=1em,xrightmargin=1em, aboveskip=1em, %设置边距
        tabsize=4, %设置tab空格数
        showspaces=false %不显示空格
       }
\usepackage{xcolor}

% \geometry{a4paper,left=2cm,right=2cm,top=1cm,bottom=1cm}

\begin{document}

\maketitle

\section{Supplementary Methods}\label{sec:suppmethods}

\subsection{Comparison to previous methods}

In order to generate background regions for hypothesis
testing of association analysis, there are two general categories of methods.  
One strategy is to sample from a larger experimental pool or
database.
Methods following this strategy include
LOLA \citep{sheffield2016lola} using Fisher's exact test, and
Poly-Enrich \citep{lee2020poly} using a likelihood ratio test based on
a Negative Binomial likelihood.
\mike{should we mention matchRanges here?}
Another strategy is to permute or shuffle the genomic regions, possibly considering an
exclusion list of regions where the original region set should not be
re-located. Example methods belonging to this category include
bedtools shuffleBed \citep{quinlan2010bedtools}, ChIP-Enrich
\citep{welch2014chip}, and
GenometriCorr \citep{GenometriCorrfavorov2012}.
In addition, the method GAT allows controlling for GC content \citep{GAT_2013},
and regioneR implements a circular shift to
preserve the clumping property of genomic regions \citep{gel2016regioner}.
Our method falls in the second category in that we redistribute
regions along the genome, though we use a block resampling scheme 
to preserve local genome structure, as proposed for genomic region sets by
\citet{bickel2010subsampling}.

\subsection{Segmentation}

In the following, we define genome segmentation and how it pertains to
block bootstrap resampling, as proposed by \citet{bickel2010subsampling}.
The overall motivation for bootstrapping with respect to a segmented
genome is to preserve large scale genomic structure, e.g. large
regions with low or high density of genomic features such as
genes. This preservation of structure through segmentation is in
addition to the preservation of local clumping of features, which is
accomplished by resampling blocks.

In this work, we segmented the genome based on gene density. We
downloaded the Ensembl genes \citep{ensemblTrimmedEntry} \mike{which
  version}, and counted the 
number of genes per million base pairs. We then supplied this vector
of counts to various methods for segmentation. We sought to define
sections of the genome that exhibit stationarity, where in this case,
stationary means similar gene density within a segment. We
additionally sought to group together segments across the genome that
had similar gene density, such that a segmentation state
$i \in [1,\dots,S]$ consisted of one or more non-contiguous regions of
the genome with similar gene density.

For block bootstrapping with respect to genome segmentations,
we considered various genome segmentation methods,
Circular Binary Search (CBS) \citep{cbs} and a hidden Markov model (HMM)
\citep{rcpphmm}, as well as pre-defined segmentations,
ChromHMM applied to 
Roadmap Epigenomics data \citep{ernst2012chromhmm}.
CBS and HMM
have R package implementations, and so we incorporated these
into a utility function in \nullranges, \texttt{segmentDensity()}.
CBS, implemented in the \textit{DNAcopy} package, was used to recursively split
chromosomes into subsegments with similar density, in this case based
on gene count per megabase. We then used k-means to cluster
chromosome subsegments into groups with similar gene
density. Typically this would result in genome segmentations with
about three states, corresponding to low, medium, and high gene
density. 
HMM was used to model the genes per megabase as an emission
density \mike{how many states}. We then used the
Viterbi algorithm for the hidden state decoding. \mike{can you talk
about the normalization of the gene counts used to run the HMM?}
Roadmap segmentations derived by ChromHMM \citep{ernst2012chromhmm} were
downloaded from
\url{https://egg2.wustl.edu/roadmap/web_portal/}. ChromHMM is based on
a multivariate Hidden Markov Model that explicitly models the presence
or absence of many chromatin marks. The original ChromHMM annotation
generated segmentation including 
15 small states. We then summarized these into 3 general categories: low
density(`E9',`E13',`E14',`E15'), middle density(`E10',`E11',`E12'),
and high density(`E1-E8'). 
After merging, our version of ChromHMM applied to Roadmap data had
8,797 ranges were left and the mean width was around 
0.33 Mb.

\subsection{Simple region shuffling}\label{sec:shuffle}

Simple region shuffling was performed by placing regions of interest
uniformly in acceptance regions with probability proportional to the
original SNP count per chromosome. For the paper, we defined
acceptance regions as the inverse of
ENCODE excluded regions \mike{cite Kundaje paper for this, see
Mikhail's vignette for which one to cite},
centromeres, and telomeres from UCSC,
as provided in the \code{excluderanges} Bioconductor package
\citep{excluderanges}.

\subsection{Assessment of \bootranges}

Suppose we are interested in computing and assessing the significance
of the overlap between one feature
\mike{set of regions? are we not using feature anymore?}
$\bm{x}$, denoted by $T_1, ..., T_\alpha$ with lengths $\tau_1,...,
\tau_\alpha$, and the feature $\bm{y}$, denoted by $Q_1, ..., Q_\beta$
with lengths $\rho_1, ..., \rho_\beta$.

Then, the region overlap is
defined as $s_{obs} \equiv \frac{1}{\alpha}\sum_{t=1}^\alpha V_t$,
where $V_t=1-\prod_{k\in (T_t+\theta)}(1-J_k)$ and $J_k=1$ if position
$k$ belongs to feature $\bm{y}$ and 0 otherwise.

$\theta$ is the
shift among feature $\bm{x}$ and $\bm{y}$, such as 1kb when linking
promoters to gene.

After block bootstrapping, we generate $R$ new
sets of null ranges $\bm{y'}$ within $R\times \frac{n}{L_b}$ blocks.

Therefore, we can do both genome-wise or block-wise analysis based on
the question to be addressed.

Then, we could use $s_{1}, s_{2}, ...,
s_{R}$ to derive genome-wise empirical \textit{p}-value
$= \frac{1}{R} \sum_{r=1}^R \mathbb{I}_{\{s_r > s_\text{obs}\}}$ or
$s_{1}, s_{2},..., s_{R\times \frac{n}{L_b}}$
to derive block-wise empirical
\textit{p}-value
$= \frac{1}{R\times \frac{n}{L_b}} \sum_{r=1}^{R\times \frac{n}{L_b}} \mathbb{I}_{\{s_r > s_{obs}\}}$,
respectively.

Additionally, we could use $z$ score to measure the
distance between the observed statistics and the bootstrap statistics
distribution in terms of standard deviations (SD).

Specifically, the
$z$ score in \cref{fig:result} is calculated by

$$ z = \frac{s_\text{obs} - \widehat{s}}{ \text{SD}_R(\widehat{s})},$$

where $\widehat{s} = \frac{\sum_{r=1}^R \widehat{s_r}}{R}$ is the sample
mean of the $R$ replications and
$\text{SD}_R(\widehat{s}) = \sqrt{\frac{\sum_{r=1}^R [\widehat{s_r}-\widehat{s}]^2}{R-1}}$.
Note that, if block-wise analysis is preferred, the SD of bootstrap
distribution should be scaled by $\sqrt{L_b}$.

\subsection{Swapping algorithm}\label{sec:algorithm}

\begin{algorithm}[H]
  \SetAlgoLined
  \KwData{Regions ($\bm{y}$), Block length ($L_b$), Length of
    chromosome $c$ ($L_c$), Total number of chromosomes ($C$),
    Bootstrap iterations ($R$), Type ($permute$ or $bootstrap$),
    Excluded regions ($\bm{e}$)}
  \KwResult{$\texttt{bootRanges}$ object, contains the bootstrapped
    region sets from each iteration, concatenated across $R$ total
    iterations. Iteration and $L_b$ recorded as metadata}
  $r \gets 1$\\
  \While{$r \leq R$}{
    \texttt{rearranged blocks} $\gets$ Generate consecutive tiling
    blocks with width = $L_b$ (except last tile per chromosome which
    is cut to fit $L_c$ for chromosome $c$)\\
    \uIf{permutation}{
\texttt{random blocks} $\gets$ Sample blocks without replacement from rearranged blocks
}\ElseIf{bootstrap}{ 
         $n_b \gets \sum_{c=1}^{C} \text{ceiling}(L_c / L_b)$\\
         \texttt{random blocks} $\gets$ Sample $n_b$ blocks with
         replacement from the genome, where sampling probability per
         chromosome is proportional to $L_c$\\
      }
\texttt{bootstrap[r]} $\gets$ Shift regions in $\bm{y}$ that fall in the \texttt{random blocks} to their
respective positions in the \texttt{rearranged blocks}\\
\texttt{bootstrap[r]} $\gets$ Trimming applied to any
regions in \texttt{bootstrap[r]} that extend past chromosome end or into
excluded regions $\bm{e}$
}
\caption{Un-segmented block bootstrap / permutation} \label{alg:supp_unseg}
\end{algorithm}

\begin{algorithm}
\SetNoFillComment
\caption{Segmented block bootstrap with proportional block length}\label{alg:framework}
  \KwData{Feature GRanges($\bm{y}$), Block length ($L_b$), Length of total genome ($L_C$), Bootstrap times ($R$), Segmentation GRanges with $L_j$ represents each ranges width and $\alpha_i$ represents ranges belong to state $i$}
  \KwResult{$\bm{bootRanges}$ object, a GRanges object with all the ranges concatenated, and iteration and block length indicated by metadata columns}
\While{$r \leq R$}{
\For{each segmentation state $i$}{
        $L_s^i = \sum_{j\in \alpha_i} L_j$ \\
         $L_b^i = L_b * L_s^i / L_C$ \\
         $n_b^i = \sum_{j\in \alpha_i} \text{ceiling}(L_{j} / L_b^i)$ \\
	\textbf{rearranged blocks:} Generate $n_b^i$ tiling blocks start site \\
         \textbf{random blocks:} Generate $n_b^i$ blocks start site with replacement and and number of blocks per ranges is proportional to $L_j$ \\
         \textbf{Return:} rearranged and random  block start and chromosome names
    }
Construct random blocks Granges\\
Move features in $\bm{y}$ that fall into random blocks to rearranged blocks given the shift as difference between random blocks and rearranged blocks start sites
}
\end{algorithm}

\begin{algorithm}
\SetNoFillComment
\caption{Segmented block bootstrap with fixed block length across chromosome}\label{alg:supp_fixedlb}
  \KwData{Feature GRanges($\bm{y}$), Block length ($L_b$), Length of total genome ($L_C$), Bootstrap times ($R$), Segmentation GRanges with $L_j$ represents each ranges width and $\beta$ represents total number of ranges}
  \KwResult{$\bm{bootRanges}$ object, a GRanges object with all the ranges concatenated, and iteration and block length indicated by metadata columns}
\While{$r \leq R$}{
         $n_{j} = \text{ceiling}(L_{j} / L_b)$ \\
         $n_b = \sum_{j=1}^{\beta} n_j$ \\
         \textbf{rearranges blocks:} Generate $n_b$ tiling blocks start site by order with width = $L_b$  \\
 	\textbf{random blocks:} Generate $n_b$ blocks start site with replacement and number of blocks per ranges is proportional to $L_j$ \\
\For{each segmentation state $i$}{
         Identify rearranged blocks that are in state $i$  \\
	Identify random blocks that are in state $i$ \\
         \textbf{Return:} random block start sites and chromosome names, rearranged block start sites and chromosome names
    }
Construct random blocks Granges \\
Move features in $\bm{y}$ that fall into random blocks to rearranged blocks given the shift as difference between random blocks and rearranged blocks start sites
}
\end{algorithm}


\section{Supplementary Results} \label{sec:results}
\subsection{Overlap analysis of liver ATAC-seq with SNPs associated with
total cholesterol}
1872 SNPs data were download from the NHGRI-EBI GWAS catalog \citep{gwascatalog} on September 22, 2021. We only extracted single variant  associated with total cholesterol. Liver ATAC-seq
\citep{CURRIN20211169} information of 20 samples were downloaded from GSE164870.
Then genomic coordinates of consensus peaks were converted from hg19 to hg38 using \code{liftOver} to construct 221,606 peaks GRanges.

\subsubsection{$L_b$ selection}\label{sec:length}
On the issue of 
block length selection, we considered it in two ways. 
One is trying to find $L_b$ that has the minimum value of a pseudo-metric $d^*(v)=|\sqrt{\frac{L_{v-1}}{L_v}}\text{IQR}(\mathcal{L}_{L_v})-\text{IQR}(\mathcal{L}_{L_{v-1}})|$, 
where $\mathcal{L}_{L_v}$ is the statistic distribution at length $L_v$, $v=1,2,\cdots,V$, $V$ is the number of candidate block lengths and $\text{IQR}(\mathcal{L})$ is the interquartile range of statistics distribution following \citet{bickel2010subsampling}. \cref{fig:suppfig0}A showed $d^*$ had smaller values in most cases when $L_b\in[300kb,800kb]$. 
Another way is evaluating conserved spatial pattern. Because we assume that generated null sets should have similar spatial properties with original sets, eg. inter-feature distance. 
We used the Earth's Mover Distance (EMD) to quantify the similarity of the distributions of an inter-feature distance between the original  and  null  models,  resulting  in  values  from  zero  (identical  distributions)  to  one  (totally  disjoint  distributions). 
The EMD between two distributions is proportional to the minimum amount of work required to change one distribution($y$) into the other ($y'$). 
Here $y$ and $y'$ are the histograms of original and null model inter-feature distance with bin size = 0.3. We observed that EMD always decreased as $L_b$ increased because more neighboring features reserved. However, $L_b$ cannot be too large and close to $L_s$. Otherwise, the blocks randomization process would not be effective. Hence, the right $L_b$ falls in the range where EMD around elbow of the line plot. In this practice, [300kb,600kb] was shown to be  a good range by visualization of density plot (\cref{fig:suppfig0}B) and according to the Elbow Method of EMD (\cref{fig:suppfig0}C). 


\subsection{Enrichment analysis of ATAC-seq and gene expressions}\label{sec:splines}
Processed 24 macrophage samples' RNA-seq  and 145 samples' ATAC-seq were loaded from \code{fluentGenomics} \citep{lee2020fluent}. They were measured after interferon gamma (IFNg) stimulation. Since the transcriptomic response to IFNg stimulation may be mediated through an increasing transcription factors binding on nearby regions and ATAC-seq can captured those regions' accessibility, we expect there would have an enrichment of differentially accessible (DA) ATAC-seq peaks in the vicinity of differentially expressed (DE) genes. 

When performing block bootstrap with 500kb block length on 100 times, we got $z_{99}=-108.1$ and $\textit{p}$-value $<.05$. We rejected the null hypothesis and concluded that there was significant enrichment of DA peaks near DE genes by counting overlaps.

For fitting the generalized penalized splines, \code{gam} function in the \code{mgcv} package was used to fit the model, which was based on a penalized likelihood maximization, and generalized cross-validation was utilized to choose the optimal value for the smoothing parameter, $\lambda$. Then, \code{tidymv} package was implemented to predict and extract the fitted value. 

\begin{lstlisting}[language=R]
boot_stats <- x %>% plyranges::join_overlap_inner(bootRanges) %>%
  plyranges::group_by(id.x,iter) %>%
  plyranges::summarize(count = plyranges::n(), logFC = max(logFC)) %>%
  as.data.frame() %>%
  tidyr::complete(`id.x`, iter, fill=list(count = 0)) %>%
  dplyr::select(iter, count, logFC) %>%
  tidyr::nest(-iter) %>%
  dplyr::mutate(
	 fit= map(data, ~gam(count ~ s(logFC), data = ., family=poisson)),
         pred  = map(fit, ~predict_gam(model = ., length_out = 2000)),
         fitted = map(pred,~find_fit(data=., logFC = seq(-8,10,1))))

\end{lstlisting} 

\subsection{Correlation analysis of Chromium Single Cell Multiome ATAC + Gene Expression assay}
Data were downloaded according to \citet{Vignette} which includes genes and peaks in 10,032 cells. 
Cell type annotations have been done as a priori by the 10x Genomics R\&D team. 
Then, information on chromosome 1 to 22 were selected to construct gene and peak GRanges. 
Since the main goal is not to accurately find gene-promoter pairs but the realization, the following preprocess may not be the most suitable way. 
First, we aggregated cells within same cell types, to form ‘pseudobulks’ with 14 samples according to the metadata because pseudo-bulk provided smoother correlation statistics without loss of the information of interest. 
Next, we removed all the features with 0 standard deviation. 
Then, log counts per million (CPM) were computed from \code{edgeR} to account for different library size.

Since genes' expression is most \textit{cis}-regulated by chromatin accessibility, there is a belief that two modalities would have significant high correlation. For the whole gene set, the mean correlation of genes and ATAC-seq read counts was 0.33, while the subsampling correlation distribution in \cref{fig:suppfig3} A had mean 0.007 across 1000 times block bootstraps. As expected,
RNA and ATAC measured at local peaks had similar cell-type-specificity.
Additionally, the average gene-peaks correlation per gene can be
computed and compared to a bootstrap distribution to
identify 5644 gene-promoter pairs that were significantly correlated across cell types, among which 5591 genes had only
one pair, 25 genes had 2 pairs. 
% The most negative correlation gene TET3 has $\rho = -0.963$\cref{fig:suppfig}G outside the subsampling confidence interval [-0.85,0.78] and previous studies have assessed down-regulation TET3 are essential for B-cell development and tumorigenesis\citep{lio2019dysregulation}. Similarly, CD83\cref{fig:suppfig}H with $\rho = 0.992$ outside [-0.80, 0.68] was a well known myeloid markers\citep{li2019cd83}. 
Those significant pairs could provide important insights into perturbation experiments for validation. Two examples were shown in \cref{fig:suppfig3}B-C. The block below shows example code for running analysis.
\begin{lstlisting}[language=R]
# split sparse count matrix into NumericList
x <- x_Granges %>%
  mutate(counts_X = NumericList(asplit(x.cpm, 1)))%>% sort()
y <- y_Granges %>%
  mutate(counts_y = NumericList(asplit(y.cpm, 1))) %>% sort()
# First standardize read counts for fast correlation computation
x$counts1 <- NumericList(lapply(x$counts_X,function(z)(z-mean(z))/sd(z)))
y$counts2 <- NumericList(lapply(y$counts_y,function(z)(z-mean(z))/sd(z)))

bootranges <- bootRanges(y,blockLength = 5e5, R=100)
  
# for standardized x and y:
correlation = function(x,y) 1/(length(x)-1) * sum(x*y)
## extract bootstrap summary statistics
boot_stats<-x %>% plyranges::join_overlap_inner(boots, maxgap=1000) %>%
  plyranges::mutate(rho = correlation(counts_x, counts_y)) %>%
  plyranges::group_by(iter) %>%
  plyranges::summarise(meanCor = mean(rho)) 
\end{lstlisting} 
 
\newpage
\section{Supplementary Figures}
\begin{figure*}[htbp]
\centering
\includegraphics[scale=0.35]{Figures/sfig1.jpeg}
\caption{$L_b$ selection assessment. A) A pseudo-metric $d^*$ over $L_b$, B) The null sets' log2(inter-feature distance+1) density plots over various $L_b$ and segmentation settings. Red line represents observed feature set's log2(inter-feature distance+1) distance. The more null sets' density plots overlapped with observed features, the better conversing spatial distribution of original set captured by block bootstrapped. Median EMD were shown as text in each panel. C) Median EMD over $L_b$ where EMD quantified similarity of inter-feature distance distributions between nulls sets and observed sets.} 
\label{fig:suppfig0}
\end{figure*}

\begin{figure*}[htbp]
\centering
\includegraphics[scale=0.3]{Figures/zscore.jpeg}
\caption{ A) \textit{z} score over -log10 (\textit{p}-value) of the liver dataset. B) \textit{z} score over gene logFC of the macrophage dataset. \textit{z} score indicated the amount of standard deviations that observed fitted overlap count away from 1000 times block bootstrap's fitted overlap count.} 
\label{fig:suppfig2}
\end{figure*}

\begin{figure*}[htbp]
\centering
\includegraphics[scale=0.08]{Figures/sfig2.jpeg}
\caption{Results of correlation analysis of Chromium Single Cell Multiome ATAC + Gene Expression assay. A) The mean correlation distribution of genes expression with 1000 times block bootstrapped ATAC's read counts. B) Gene TET3 read counts over peak chr2:74000098-74003475 read counts, colored by cell types. TET3 has the most negative correlation $\rho = −0.963$. C) Gene CD83 read counts over peak chr6:14116971-14139988 read counts, colored by cell types. The correlation of this gene-promoter pair is 0.992.} 
\label{fig:suppfig3}
\end{figure*}

\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}
%
%\bibliographystyle{plain}
%
\bibliography{document}

\end{document}
